// storybook-addon-performance/dist/panel/task-result/timed-result.js
import React from 'react';
import toFixed from '../../util/to-fixed';
import { ExpandingResult } from './expanding-result';
import * as Parts from './parts';
const warningIcon = React.createElement("span", null, "\u26A0\uFE0F");
function getDiff({ result, pinned }) {
    if (!pinned) {
        return 0;
    }
    return toFixed(1 - pinned.averageMs / result.averageMs);
}
function DiffToPinned({ result, pinned }) {
    if (pinned == null) {
        return null;
    }
    return (React.createElement(React.Fragment, null,
        React.createElement(Parts.Heading, null, "Compared with pinned"),
        React.createElement(Parts.Content, null,
            React.createElement(Parts.Table, null,
                React.createElement("tbody", null,
                    React.createElement("tr", null,
                        React.createElement(Parts.TitleCell, null, "Pinned value"),
                        React.createElement(Parts.ValueCell, null,
                            React.createElement(Parts.ValueLozenge, { type: "raw" },
                                toFixed(pinned.averageMs, 1),
                                "ms"))),
                    React.createElement("tr", null,
                        React.createElement(Parts.TitleCell, null, "Current value"),
                        React.createElement(Parts.ValueCell, null,
                            React.createElement(Parts.ValueLozenge, { type: "raw" },
                                toFixed(result.averageMs, 1),
                                "ms"))),
                    React.createElement("tr", null,
                        React.createElement(Parts.TitleCell, null, "Difference"),
                        React.createElement(Parts.ValueCell, null,
                            React.createElement(DiffLozenge, { diff: getDiff({ result, pinned }) }))))))));
}
function Variance({ result }) {
    if (result.samples === 1) {
        return null;
    }
    const wasStable = result.variance.standardDeviation < 1;
    return (React.createElement(React.Fragment, null,
        React.createElement(Parts.Heading, null, "Variance"),
        React.createElement(Parts.Note, null, "When doing multiple runs the can be differences between the runs. The lower the variance, the higher confidence you can have"),
        React.createElement(Parts.Table, null,
            React.createElement("tbody", null,
                React.createElement("tr", null,
                    React.createElement(Parts.TitleCell, null, "Samples"),
                    React.createElement(Parts.ValueCell, null,
                        React.createElement(Parts.ValueLozenge, { type: "info" }, result.samples))),
                React.createElement("tr", null,
                    React.createElement(Parts.TitleCell, null, "Standard deviation"),
                    React.createElement(Parts.ValueCell, null,
                        React.createElement(Parts.ValueLozenge, { type: wasStable ? 'positive' : 'negative' },
                            wasStable ? null : warningIcon,
                            toFixed(result.variance.standardDeviation)))),
                React.createElement("tr", null,
                    React.createElement(Parts.TitleCell, null, "Lowest"),
                    React.createElement(Parts.ValueCell, null,
                        React.createElement(Parts.ValueLozenge, { type: "raw" },
                            "-",
                            toFixed(result.variance.lowerPercentage),
                            "%"))),
                React.createElement("tr", null,
                    React.createElement(Parts.TitleCell, null, "Highest"),
                    React.createElement(Parts.ValueCell, null,
                        React.createElement(Parts.ValueLozenge, { type: "raw" },
                            "+",
                            toFixed(result.variance.upperPercentage),
                            "%")))))));
}
function Expanded({ task, result, pinned, }) {
    return (React.createElement(React.Fragment, null,
        React.createElement(DiffToPinned, { result: result, pinned: pinned }),
        React.createElement(Variance, { result: result }),
        React.createElement(Parts.Heading, null, "Description"),
        React.createElement(Parts.Content, null, task.description)));
}
function DiffLozenge({ diff }) {
    const type = (() => {
        if (diff > 1) {
            return 'negative';
        }
        if (diff < -1) {
            return 'positive';
        }
        return 'faint';
    })();
    return React.createElement(Parts.ValueLozenge, { type: type },
        diff,
        "%");
}
export default function TimedResultView({ task, pinned, result }) {
    const diff = getDiff({ result, pinned });
    const resultNode = (React.createElement(React.Fragment, null,
        diff ? React.createElement(DiffLozenge, { diff: diff }) : null,
        React.createElement(Parts.ResultValue, null, toFixed(result.averageMs)),
        ' ',
        React.createElement(Parts.ResultScale, null, "ms")));
    return (React.createElement(ExpandingResult, { taskId: task.taskId, name: task.name, result: resultNode, getExpanded: ({ isExpanded }) => isExpanded ? React.createElement(Expanded, { task: task, result: result, pinned: pinned }) : null }));
}
